pulsar-io/jdbc/src/main/java/org/apache/pulsar/io/jdbc/JdbcAbstractSink.java

```diff
@@ -0,0 +1,190 @@
+        flushExecutor.scheduleAtFixedRate(() -> flush(), timeoutMs, timeoutMs, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (!connection.getAutoCommit()) {
+            connection.commit();
+        }
+        flushExecutor.shutdown();
+        if (connection != null) {
+            connection.close();
+        }
+        log.info("Connection Closed");
+    }
+
+    @Override
+    public void write(Record<T> record) throws Exception {
+        int number;
+        synchronized (incomingList) {
+            incomingList.add(record);
+            number = incomingList.size();
+        }
+
+        if (number == batchSize) {
+            flushExecutor.schedule(() -> flush(), 0, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    // bind value with a PreparedStetement
+    public abstract void bindValue(
+        PreparedStatement statement,
+        Record<T> message) throws Exception;
+
+
+    private void flush() {
+        // if not in flushing state, do flush, else return;
+        if (isFlushing.compareAndSet(false, true)) {
+            log.debug("Starting flush, queue size: {}", incomingList.size());
+            checkState(swapList.isEmpty(),
+                "swapList should be empty since last flush. swapList.size: " + swapList.size());
+
+            synchronized (incomingList) {
+                List<Record<T>> tmpList;
+                swapList.clear();
+
+                tmpList = swapList;
+                swapList = incomingList;
+                incomingList = tmpList;
+            }
+
+            int updateCount = 0;
+            boolean noInfo = false;
+            try {
+                // bind each record value
+                for (Record<T> record : swapList) {
+                    bindValue(insertStatement, record);
+                    insertStatement.addBatch();
+                    record.ack();
+                }
+
+                for (int updates : insertStatement.executeBatch()) {
+                    if (updates == Statement.SUCCESS_NO_INFO) {
+                        noInfo = true;
+                        continue;
+                    }
+                    updateCount += updateCount;
+                }
+                if (!connection.getAutoCommit()) {
+                    connection.commit();
+                }
+            } catch (Exception e) {
+                log.error("Got exception ", e);
+                swapList.forEach(tRecord -> tRecord.fail());
+            }
+
+            if (swapList.size() != updateCount) {
+                log.error("Update count {}  not match total number of records {}", updateCount, swapList.size());
+            }
+
+            // finish flush
+            log.debug("Finish flush, queue size: {}", swapList.size());
+            isFlushing.set(false);
+        } else {
+            log.debug("Already in flushing state, will not flush, queue size: {}", incomingList.size());
```

Comment:
if (log.isDebugEnabled())

Reply:
incomingList should be protecred under a synchronization block

Reply:
Thanks， This is just to get info, in log, no operation on the list.